### ISSUE TITLE
ReLU layer wrong result with negative threshold

### ISSUE THREAD

**Issue by rschumi0** [tensorflow/tensorflow#48646]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.4.0-rc4-71-g582c8d236cb 2.4.0
- Python version: 3.8.5

**Describe the current behavior**
The [ReLu](https://www.tensorflow.org/api_docs/python/tf/keras/layers/ReLU) layer seems to have an issue with negative thresholds, even though it should support them according to the documentation. For example, for a  `max_value=1` and `threshold=-1`, ReLu should produce f(x)=0.5 for x=0.5 according to the following formula from the documentation  `f(x) = x if threshold <= x < max_value`. 
The issue is illustrated in the following code.

**Standalone code to reproduce the issue**
```
import tensorflow as tf
from tensorflow import keras
model = keras.Sequential([
keras.layers.ReLU(max_value=1, threshold=-1, negative_slope=1, input_shape=(4,))])
x = tf.constant([[1.5, 0.5,-0.5, -1.5]])
print (model.predict(x,steps=1))
```
`Output: [[ 1.   0.5  0.  -0.5]] Expected Output: [[1   0.5  -0.5  -0.5]]`

**Comment by szutenberg** [comment]
That's how "ReLU" looks like if we obey 
```
  f(x) = max_value if x >= max_value
  f(x) = x if threshold <= x < max_value
  f(x) = negative_slope * (x - threshold) otherwise

negative_slope = 1
threshold = -1
max_value = 1
```
![image](https://user-images.githubusercontent.com/37601244/115468201-07091680-a233-11eb-9567-796d263a6ad5.png)

It does not look like "**Re**ctified **L**inear **U**nit anymore so I think it's no point to fix it and I recommend modifying the docs. (https://github.com/tensorflow/tensorflow/pull/48654)

**Comment by AdityaKane2001** [comment]
>   f(x) = negative_slope * (x - threshold) otherwise

Can we do
` f(x) =   negative_slope * (x - threshold) + threshold `

This will work for the general case as `f(x) .= x for x = threshold`

**Comment by AdityaKane2001** [comment]
@szutenberg 
Also, there's something more weird going on with `tf.nn.relu` function. Here's how it works:
```python
a = np.array( [ [0.5,2] , [-1.5,-0.5] ])
# array([[ 0.5,  2. ],
#       [-1.5, -0.5]])
print(tf.keras.backend.relu(a, alpha = 1, max_value = 1, threshold=-1)) 
# Output :
# tf.Tensor(
# [[ 0.5  1. ]
# [-0.5  0. ]], shape=(2, 2), dtype=float64)
```

According to your graph the (1,1) entry should be -0.5.

**Comment by szutenberg** [comment]
@AdityaKane2001 

Yes, it does not match my graph because I drew it by following:
```
  f(x) = max_value if x >= max_value
  f(x) = x if threshold <= x < max_value
  f(x) = negative_slope * (x - threshold) otherwise
```

That's why I propose PR https://github.com/tensorflow/tensorflow/pull/48654 which solves the problem (negative threshold is simply not supported).

I tried to understand why the threshold parameter was introduced. I found commit https://github.com/tensorflow/tensorflow/commit/0cf2c612e5e6ff8c5026011e8186056801def747 . I don't understand what is the point to introduce such consolidation. Any ideas?

Note that in [keras.layers.ThresholdedReLU](https://www.tensorflow.org/api_docs/python/tf/keras/layers/ThresholdedReLU) `theta >= 0` (`theta` is equivalent to `threshold` in ReLU).

We should not replace `f(x) = negative_slope * (x - threshold)` with `f(x) = negative_slope * (x - threshold) + threshold` because it would break compatibility with [keras.layers.ThresholdedReLU](https://www.tensorflow.org/api_docs/python/tf/keras/layers/ThresholdedReLU) by modifying behaviour for positive theshold value (f(x) would return threshold instead of 0).

**Comment by AdityaKane2001** [comment]
@szutenberg 
Fair enough. But then should we modify the relu function itself, asserting that we do have a non-negative threshold? It will remove any ambiguity for the user in case the user tries to enter a negative threshold value.

**Comment by szutenberg** [comment]
@AdityaKane2001 I added additional asserts to ReLU.

From the graph I drew, we can see that such scenario **probably** does not make any sense. It's not continuous at -1, this sudden change may break gradient descent.

I don't understand what is the idea behind https://github.com/tensorflow/tensorflow/commit/0cf2c612e5e6ff8c5026011e8186056801def747 and I'd rather revert it than eliminate ThresholdedReLU. People use ThresholdedReLU: [example 1](https://github.com/gao-lab/REVA-Data_Source_Code/blob/9da1a9fcfa04663b6a09a9c2549af9dddcc9848a/Variant_annotation/CNNs/model_structures.py#L17), [example 2](https://github.com/tincochan/vGame_bgm_remix/blob/4135b1a5ff9f1de9107facb8623d586200c0246e/rnn-cnn-gan-enhancer.py#L66)

For very rare use cases users can implement their own layers.

**Comment by AdityaKane2001** [comment]
@szutenberg 
I agree with you. The use case is very (very) small..
